Sources/PerformanceMetrics/PerformanceMetrics.swift

```diff
@@ -1,52 +1,124 @@
-        trackers.forEach { $0.begin(with: identifier) }
-    }
-    
-    public func end(with identifier: PerformanceMetricsTracker.Identifier, metadata: PerformanceMetricsTracker.Metadata? = nil) {
-        trackers.forEach { $0.end(with: identifier, metadata: metadata) }
+            assert(subTokens.count == trackers.count, "😱: number of sub tokens and sub trackers must match!")
+
+            zip(trackers, subTokens).forEach { (tracker, token) in tracker.stop(with: token, metadata: metadata) }
+        }
+
+        /// Measures a given closure's execution time once it returns or throws (i.e. *synchronously*). An optional
+        /// metadata dictionary can be provided to be associated with the recorded metric. The metric is calculated on
+        /// all sub trackers.
+        ///
+        /// - Parameters:
+        ///   - identifier: The metric's identifier.
+        ///   - metadata: The metric's metadata dictionary. The default is `nil`.
+        ///   - execute: The closure to measure the execution time of.
+        /// - Returns: The closure's return value, if any.
+        /// - Throws: The closure's thrown error, if any.
+        @discardableResult
+        public func measure<T>(with identifier: Identifier,
+                               metadata: Metadata? = nil,
+                               execute: () throws -> T) rethrows -> T {
+
+            let subTokens = trackers.map { $0.start(with: identifier) }
+            defer {
+                zip(trackers, subTokens).forEach { (tracker, token) in tracker.stop(with: token, metadata: metadata) }
+            }
+
+            let measureResult = try execute()
+
+            return measureResult
+        }
+
+        /// Measures a given closure's execution time *until* an inner `stop` closure is invoked by it with an optional
+        /// metadata dictionary to be associated with the recorded metric. The metric is calculated on all sub trackers.
+        ///
+        /// Should be used for asynchronous code, or when the metric metadata is only available during execution.
+        ///
+        /// - Important: The `stop` closure should be called *before* the `execute` either returns or throws.
+        ///
+        /// - Parameters:
+        ///   - identifier: The metric's identifier.
+        ///   - execute: The closure to measure the execution time of.
+        ///   - stop: The closure to be invoked by `execute` to stop measuring the execution time, along with any
+        /// additional metric metadata.
+        ///   - metadata: The metric's metadata dictionary.
+        /// - Returns: The closure's return value, if any.
+        /// - Throws: The closure's thrown error, if any.
+        @discardableResult
+        public func measure<T>(with identifier: Identifier,
+                               execute: (_ stop: (_ metadata: Metadata?) -> Void) throws -> T) rethrows -> T {
+
+            let subTokens = trackers.map { $0.start(with: identifier) }
```

Comment:
We repeat this logic to create the subtokens 3 times in this file. Is it worth to move this to a function?

Reply:
I didn't do it since it's such a simple logic (and copy paste is super efficient these days 😇)